File: src/tauri/configuration.ts
import { invoke } from "@tauri-apps/api";
import { addNotification, AppState } from "../providers";
import { format, TauriError } from "./error";

export const getConfiguration = (): Promise<AppState> =>
  invoke<AppState>("get_configuration").catch((err: TauriError) => {
    addNotification({ type: "error", title: "Unable to retrieve the user config", description: format(err) });
    throw err;
  });

export const setConfiguration = (configuration: AppState): Promise<AppState> =>
  invoke<AppState>("write_configuration", { configuration }).catch((err: TauriError) => {
    addNotification({ type: "error", title: "Unable to update the user config", description: format(err) });
    throw err;
  });

File: backend/src/api/configuration.rs
use log::debug;

use crate::core::configuration::InsulatorConfig;

use super::{error::ApiResult, AppState};

#[tauri::command]
pub fn get_configuration(state: tauri::State<'_, AppState>) -> ApiResult<InsulatorConfig> {
    debug!("Retrieve configuration");
    Ok(state.configuration_provider.get_configuration()?)
}

#[tauri::command]
pub fn write_configuration(
    configuration: InsulatorConfig,
    state: tauri::State<'_, AppState>,
) -> ApiResult<InsulatorConfig> {
    debug!("Write configuration");
    Ok(state
        .configuration_provider
        .write_configuration(&configuration)
        .map(|_| configuration)?)
}
